## Overview

In this PA, you will will be creating and manipulating basic objects to uncover a top secret message.

### Hint 1

As you work through this PA, you will encounter some code that does not work as you want it to.  Don't despair!  Errors (when R is unable to run your code) and bugs (when the code doesn't do what you hoped) are a natural part of coding.  Even the best of the best deal with these problems regularly - learning to track down the issue is a skill that you can learn and practice.

Google is your friend!  If an error is confusing, copy it into Google and see what other people are saying.  If you don't know how to do something, search for it.

### Hint 2

Just because there is no error message doesn't mean everything went smoothly.  You should be regularly printing out your objects, to see they contain the information and format that you expect.

For example, if you have an object named `my_thing`, you might first want to simply look at the object by typing its name.

However, you might also want to make sure it is the right **object type** and/or the right **object structure**.


In [ ]:
type(my_thing)
len(my_thing)

NameError: name 'my_thing' is not defined

In [2]:
# Define a character vector

from numpy import array
weekdays = array(["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"])
weekend = array(["Sunday", "Saturday"])
# Create logical vectors manually
relax_days = array([True, False, False, False, False, False, True])

# # Create logical vectors automatically
from numpy import isin     # get a special function for arrays
relax_days = isin(weekdays, weekend)

relax_days

# # Using logical vectors to index the character vector
weekdays[relax_days]


# # Using ~ to reverse the True and False
weekdays[~relax_days]

array(['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday'],
      dtype='<U9')

## Part One: Setup

Each of the following code chunks will cause an error and/or do the desired task incorrectly.  Find the mistake, and correct it to complete the intended action.




1. Create lists containing the upper case letters, lower case letters, and some punctuation marks.


In [3]:
lower_case = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"]

upper_case = ["A", "B", "C", "D", "E", "F", "G", "H" "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"]

punctuation = [".", ",", "!", "?", "'", '"', "(", ")", " ", "-", ";", ":"]

2. Make one long list containing all the symbols.  Find the length of this list.


In [4]:
all_symbols = lower_case + upper_case + punctuation

len(all_symbols)

63

3. Repeat the previous steps, but instead of using *list* type objects, use **numpy arrays**.  Don't forget to **import** the numpy library!


In [5]:
import numpy as np

lower_case = np.array(["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"])

upper_case = np.array(["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"])

punctuation = np.array([".", ",", "!", "?", "'", "\"", "(", ")", " ", "-", ";", ":"])

## Use np.concatenate to combine into long array
all_symbols = np.concatenate((lower_case, upper_case, punctuation))
all_symbols

array(['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm',
       'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
       'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M',
       'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z',
       '.', ',', '!', '?', "'", '"', '(', ')', ' ', '-', ';', ':'],
      dtype='<U1')

4. Make a **pandas data frame**, with one column called "Symbol" containing the `all_symbols` array.


In [6]:
import pandas as pd

dat = pd.read_csv("https://gist.githubusercontent.com/slopp/ce3b90b9168f2f921784de84fa445651/raw/4ecf3041f0ed4913e7c230758733948bc561f434/penguins.csv")

dat
dat.body_mass_g
#dat.info()

0      3750.0
1      3800.0
2      3250.0
3         NaN
4      3450.0
        ...  
339    4000.0
340    3400.0
341    3775.0
342    4100.0
343    3775.0
Name: body_mass_g, Length: 344, dtype: float64

In [9]:
import pandas as pd

symbols_df = pd.DataFrame(all_symbols, columns = ['Symbol'])

symbols_df

Symbol
0       a
1       b
2       c
3       d
4       e
..    ...
59      )
60       
61      -
62      ;
63      :

[64 rows x 1 columns]

## Part Two: Decoding the secret message.


In [10]:
top_secret = pd.read_csv("https://www.dropbox.com/s/k72h1zewk4gtqep/PA_Secret_Code?dl=1", header=None)

top_secret = top_secret[0].to_numpy()
top_secret


array([0.39104271, 0.38644508, 0.37131377, 0.33881179, 0.37131377,
       0.39695721, 0.35479413, 0.37573224, 0.3693288 , 0.37404822,
       0.37573224, 0.37490299, 0.37490299, 0.40018665, 0.40307943,
       0.37490299, 0.38531114, 0.38024665, 0.40542   , 0.39915133,
       0.38644508, 0.3858839 , 0.40757858, 0.39457479, 0.3924204 ,
       0.39196872, 0.40368332, 0.39330231, 0.40052335, 0.41099418,
       0.40597557, 0.39915133, 0.40214824, 0.39695721, 0.40246212,
       0.41362081, 0.41403541, 0.50672129, 0.46904194, 0.4189279 ,
       0.40513799, 0.50677976, 0.41403541, 0.46026337, 0.44516294,
       0.41006039, 0.50685267, 0.40485308, 0.43565499, 0.40982195,
       0.50691085, 0.41751216, 0.42308635, 0.43400049, 0.40933885,
       0.43181048, 0.4148465 , 0.50701239, 0.42481402, 0.43765951,
       0.50705579, 0.48348599, 0.42352845, 0.41751216, 0.50711356,
       0.45815983, 0.41403541, 0.41504562, 0.41927084, 0.50976041,
       0.50719998, 0.4658871 , 0.43527447, 0.43546538, 0.50725

Add 14 to every number.
Multiply every number by 18, then subtract 257.
Exponentiate every number. (That is, do e^[number]. You may have to Google how to do this!)
Square every number.

In [11]:
top_secret1 = top_secret + 14
top_secret2 = top_secret1*18
top_secret3 = top_secret2-257
top_secret4 = np.exp(top_secret3)
top_secret5 = top_secret4**2
top_secret5

array([5.900e+01, 5.000e+01, 2.900e+01, 9.000e+00, 2.900e+01, 7.300e+01,
       1.600e+01, 3.400e+01, 2.700e+01, 3.200e+01, 3.400e+01, 3.300e+01,
       3.300e+01, 8.200e+01, 9.100e+01, 3.300e+01, 4.800e+01, 4.000e+01,
       9.900e+01, 7.900e+01, 5.000e+01, 4.900e+01, 1.070e+02, 6.700e+01,
       6.200e+01, 6.100e+01, 9.300e+01, 6.400e+01, 8.300e+01, 1.210e+02,
       1.010e+02, 7.900e+01, 8.800e+01, 7.300e+01, 8.900e+01, 1.330e+02,
       1.350e+02, 3.797e+03, 9.780e+02, 1.610e+02, 9.800e+01, 3.805e+03,
       1.350e+02, 7.130e+02, 4.140e+02, 1.170e+02, 3.815e+03, 9.700e+01,
       2.940e+02, 1.160e+02, 3.823e+03, 1.530e+02, 1.870e+02, 2.770e+02,
       1.140e+02, 2.560e+02, 1.390e+02, 3.837e+03, 1.990e+02, 3.160e+02,
       3.843e+03, 1.645e+03, 1.900e+02, 1.530e+02, 3.851e+03, 6.610e+02,
       1.350e+02, 1.400e+02, 1.630e+02, 4.236e+03, 3.863e+03, 8.730e+02,
       2.900e+02, 2.920e+02, 3.871e+03, 3.210e+02, 2.350e+02, 3.250e+02,
       5.190e+02, 7.850e+02, 3.883e+03, 6.930e+02, 

Checkpoint: Headquarters has informed you that at this stage of decoding, there should be 352 numbers in the secret message that are below 17. Write code to check if this is true for your data.

Turn your one-dimensional array of numbers into a two-dimensional array (matrix) with 5 columns.
Separately from your top secret numbers, create an array of all the even numbers between 1 and 382. Name it "evens". That is, "evens" should contain 2, 4, 6, 8 ..., 382.
Subtract the "evens" vector from the first column of your secret message matrix.
Subtract 100 from all numbers in the 18-24th rows of the 3rd column.
Multiply all numbers in the 4th and 5th column by 2.
Turn your matrix back into a one dimensional array.

In [12]:

top_secret5_filtered = top_secret5[top_secret5 < 17]
len(top_secret5_filtered)


352

In [13]:


import numpy as np
top_secret5_matrix = np.reshape(top_secret5, (-1, 5), order = 'F')
print(top_secret5_matrix)


[[5.900e+01 1.440e+02 3.610e+02 9.000e+00 2.050e+01]
 [5.000e+01 1.000e+00 1.521e+03 7.500e+00 3.050e+01]
 [2.900e+01 5.290e+02 2.250e+02 1.050e+01 3.000e+00]
 [9.000e+00 3.610e+02 2.250e+02 3.500e+00 9.000e+00]
 [2.900e+01 3.721e+03 1.600e+01 4.000e+00 5.000e-01]
 [7.300e+01 1.521e+03 3.721e+03 3.050e+01 1.000e+00]
 [1.600e+01 6.400e+01 1.000e+00 5.000e-01 5.000e+00]
 [3.400e+01 1.000e+00 5.290e+02 7.000e+00 7.500e+00]
 [2.700e+01 1.521e+03 6.400e+01 2.000e+00 1.050e+01]
 [3.200e+01 3.721e+03 8.100e+01 3.050e+01 9.500e+00]
 [3.400e+01 9.000e+00 1.440e+02 1.950e+01 3.050e+01]
 [3.300e+01 1.000e+00 2.500e+01 4.000e+00 2.000e+00]
 [3.300e+01 1.521e+03 3.721e+03 9.000e+00 5.000e-01]
 [8.200e+01 9.000e+00 8.100e+01 7.500e+00 1.250e+01]
 [9.100e+01 6.400e+01 1.960e+02 1.050e+01 2.750e+01]
 [3.300e+01 3.025e+03 3.721e+03 3.500e+00 3.050e+01]
 [4.800e+01 3.721e+03 1.521e+03 4.000e+00 1.450e+01]
 [4.000e+01 7.840e+02 1.640e+02 3.050e+01 5.000e-01]
 [9.900e+01 2.500e+01 3.250e+02 3.050e+01 6.00

In [14]:


import numpy as np

even_numbers = np.arange(2, 383, 2)
print(even_numbers)
type (even_numbers)

[  2   4   6   8  10  12  14  16  18  20  22  24  26  28  30  32  34  36
  38  40  42  44  46  48  50  52  54  56  58  60  62  64  66  68  70  72
  74  76  78  80  82  84  86  88  90  92  94  96  98 100 102 104 106 108
 110 112 114 116 118 120 122 124 126 128 130 132 134 136 138 140 142 144
 146 148 150 152 154 156 158 160 162 164 166 168 170 172 174 176 178 180
 182 184 186 188 190 192 194 196 198 200 202 204 206 208 210 212 214 216
 218 220 222 224 226 228 230 232 234 236 238 240 242 244 246 248 250 252
 254 256 258 260 262 264 266 268 270 272 274 276 278 280 282 284 286 288
 290 292 294 296 298 300 302 304 306 308 310 312 314 316 318 320 322 324
 326 328 330 332 334 336 338 340 342 344 346 348 350 352 354 356 358 360
 362 364 366 368 370 372 374 376 378 380 382]


numpy.ndarray

In [15]:
top_secret_subtracted = top_secret5_matrix[:,0] - even_numbers
top_secret_subtracted

array([5.700e+01, 4.600e+01, 2.300e+01, 1.000e+00, 1.900e+01, 6.100e+01,
       2.000e+00, 1.800e+01, 9.000e+00, 1.200e+01, 1.200e+01, 9.000e+00,
       7.000e+00, 5.400e+01, 6.100e+01, 1.000e+00, 1.400e+01, 4.000e+00,
       6.100e+01, 3.900e+01, 8.000e+00, 5.000e+00, 6.100e+01, 1.900e+01,
       1.200e+01, 9.000e+00, 3.900e+01, 8.000e+00, 2.500e+01, 6.100e+01,
       3.900e+01, 1.500e+01, 2.200e+01, 5.000e+00, 1.900e+01, 6.100e+01,
       6.100e+01, 3.721e+03, 9.000e+02, 8.100e+01, 1.600e+01, 3.721e+03,
       4.900e+01, 6.250e+02, 3.240e+02, 2.500e+01, 3.721e+03, 1.000e+00,
       1.960e+02, 1.600e+01, 3.721e+03, 4.900e+01, 8.100e+01, 1.690e+02,
       4.000e+00, 1.440e+02, 2.500e+01, 3.721e+03, 8.100e+01, 1.960e+02,
       3.721e+03, 1.521e+03, 6.400e+01, 2.500e+01, 3.721e+03, 5.290e+02,
       1.000e+00, 4.000e+00, 2.500e+01, 4.096e+03, 3.721e+03, 7.290e+02,
       1.440e+02, 1.440e+02, 3.721e+03, 1.690e+02, 8.100e+01, 1.690e+02,
       3.610e+02, 6.250e+02, 3.721e+03, 5.290e+02, 

 Subtract 100 from all numbers in the 18-24th rows of the 3rd column. Multiply all numbers in the 4th and 5th column by 2. Turn your matrix back into a one dimensional array.

In [16]:
top_secret_6 = top_secret5_matrix[17:23,2] - 100
top_secret_6

array([  64.,  225.,  441.,   49.,   64., 1521.])

In [17]:
top_secret_7 = top_secret5_matrix[:,3:4]*2
#top_secret_7

In [18]:
top_secret5_matrix[:,0]= top_secret_subtracted
top_secret5_matrix[17:23,2]= top_secret_6
top_secret5_matrix[:,3:4]= top_secret_7
top_secret5_matrix

array([[5.700e+01, 1.440e+02, 3.610e+02, 1.800e+01, 2.050e+01],
       [4.600e+01, 1.000e+00, 1.521e+03, 1.500e+01, 3.050e+01],
       [2.300e+01, 5.290e+02, 2.250e+02, 2.100e+01, 3.000e+00],
       [1.000e+00, 3.610e+02, 2.250e+02, 7.000e+00, 9.000e+00],
       [1.900e+01, 3.721e+03, 1.600e+01, 8.000e+00, 5.000e-01],
       [6.100e+01, 1.521e+03, 3.721e+03, 6.100e+01, 1.000e+00],
       [2.000e+00, 6.400e+01, 1.000e+00, 1.000e+00, 5.000e+00],
       [1.800e+01, 1.000e+00, 5.290e+02, 1.400e+01, 7.500e+00],
       [9.000e+00, 1.521e+03, 6.400e+01, 4.000e+00, 1.050e+01],
       [1.200e+01, 3.721e+03, 8.100e+01, 6.100e+01, 9.500e+00],
       [1.200e+01, 9.000e+00, 1.440e+02, 3.900e+01, 3.050e+01],
       [9.000e+00, 1.000e+00, 2.500e+01, 8.000e+00, 2.000e+00],
       [7.000e+00, 1.521e+03, 3.721e+03, 1.800e+01, 5.000e-01],
       [5.400e+01, 9.000e+00, 8.100e+01, 1.500e+01, 1.250e+01],
       [6.100e+01, 6.400e+01, 1.960e+02, 2.100e+01, 2.750e+01],
       [1.000e+00, 3.025e+03, 3.721e+03,

In [19]:
len(top_secret5_matrix) *5

955

In [20]:
top_secret_1d = np.reshape(top_secret5_matrix, 955, order = 'F')
top_secret_1d


array([5.700e+01, 4.600e+01, 2.300e+01, 1.000e+00, 1.900e+01, 6.100e+01,
       2.000e+00, 1.800e+01, 9.000e+00, 1.200e+01, 1.200e+01, 9.000e+00,
       7.000e+00, 5.400e+01, 6.100e+01, 1.000e+00, 1.400e+01, 4.000e+00,
       6.100e+01, 3.900e+01, 8.000e+00, 5.000e+00, 6.100e+01, 1.900e+01,
       1.200e+01, 9.000e+00, 3.900e+01, 8.000e+00, 2.500e+01, 6.100e+01,
       3.900e+01, 1.500e+01, 2.200e+01, 5.000e+00, 1.900e+01, 6.100e+01,
       6.100e+01, 3.721e+03, 9.000e+02, 8.100e+01, 1.600e+01, 3.721e+03,
       4.900e+01, 6.250e+02, 3.240e+02, 2.500e+01, 3.721e+03, 1.000e+00,
       1.960e+02, 1.600e+01, 3.721e+03, 4.900e+01, 8.100e+01, 1.690e+02,
       4.000e+00, 1.440e+02, 2.500e+01, 3.721e+03, 8.100e+01, 1.960e+02,
       3.721e+03, 1.521e+03, 6.400e+01, 2.500e+01, 3.721e+03, 5.290e+02,
       1.000e+00, 4.000e+00, 2.500e+01, 4.096e+03, 3.721e+03, 7.290e+02,
       1.440e+02, 1.440e+02, 3.721e+03, 1.690e+02, 8.100e+01, 1.690e+02,
       3.610e+02, 6.250e+02, 3.721e+03, 5.290e+02, 

Checkpoint: Headquarters has informed you that at this stage of decoding, all numbers in indices 500 and beyond are below 100. Write code to check if your numbers fulfill this requirement.

Take the square root of all numbers in indices 37 to 464.
Round all numbers to the nearest whole number.
Replace all instances of the number 39 with 20.

In [21]:
len(top_secret_1d[500:])
len(top_secret_1d[500:]< 100)

455

In [22]:
top_secret_square_root = np.sqrt(top_secret_1d[37:464])
top_secret_square_root

array([61.        , 30.        ,  9.        ,  4.        , 61.        ,
        7.        , 25.        , 18.        ,  5.        , 61.        ,
        1.        , 14.        ,  4.        , 61.        ,  7.        ,
        9.        , 13.        ,  2.        , 12.        ,  5.        ,
       61.        ,  9.        , 14.        , 61.        , 39.        ,
        8.        ,  5.        , 61.        , 23.        ,  1.        ,
        2.        ,  5.        , 64.        , 61.        , 27.        ,
       12.        , 12.        , 61.        , 13.        ,  9.        ,
       13.        , 19.        , 25.        , 61.        , 23.        ,
        5.        , 18.        ,  5.        , 61.        , 39.        ,
        8.        ,  5.        , 61.        ,  2.        , 15.        ,
       18.        , 15.        ,  7.        , 15.        , 22.        ,
        5.        , 19.        , 54.        , 61.        , 61.        ,
       61.        , 27.        , 14.        ,  4.        , 61.  

In [23]:
top_secret_1d[37:464] = top_secret_square_root


In [24]:
top_secret_rounded = np.round(top_secret_1d)
len(top_secret_rounded)

955

In [25]:
# prompt: replace every instance of 39 with 20

top_secret_rounded[top_secret_rounded == 39] = 20

top_secret = top_secret_rounded
top_secret
top_secret = top_secret.astype(int)
top_secret

array([57, 46, 23,  1, 19, 61,  2, 18,  9, 12, 12,  9,  7, 54, 61,  1, 14,
        4, 61, 20,  8,  5, 61, 19, 12,  9, 20,  8, 25, 61, 20, 15, 22,  5,
       19, 61, 61, 61, 30,  9,  4, 61,  7, 25, 18,  5, 61,  1, 14,  4, 61,
        7,  9, 13,  2, 12,  5, 61,  9, 14, 61, 20,  8,  5, 61, 23,  1,  2,
        5, 64, 61, 27, 12, 12, 61, 13,  9, 13, 19, 25, 61, 23,  5, 18,  5,
       61, 20,  8,  5, 61,  2, 15, 18, 15,  7, 15, 22,  5, 19, 54, 61, 61,
       61, 27, 14,  4, 61, 20,  8,  5, 61, 13, 15, 13,  5, 61, 18,  1, 20,
        8, 19, 61, 15, 21, 20,  7, 18,  1,  2,  5, 53, 61, 58, 28,  5, 23,
        1, 18,  5, 61, 20,  8,  5, 61, 36,  1,  2,  2,  5, 18, 23, 15,  3,
       11, 54, 61, 13, 25, 61, 19, 15, 14, 55, 61, 61, 61, 46,  8,  5, 61,
       10,  1, 23, 19, 61, 20,  8,  1, 20, 61,  2,  9, 20,  5, 54, 61, 20,
        8,  5, 61,  3, 12,  1, 23, 19, 61, 20,  8,  1, 20, 61,  3,  1, 20,
        3,  8, 55, 61, 28,  5, 23,  1, 18,  5, 61, 20,  8,  5, 61, 36, 21,
        2, 10, 21,  2, 61

In [ ]:
len(top_secret)

955

In [26]:

even_count =[top_secret % 2 == 0]
np.sum(even_count)


350

### Hint

Recall that basic math operations are **vectorized**.  This means you don't have to repeat the process for each item in a list; you can apply the math to the whole thing at once.

### Instructions

By altering this top secret set of numbers, you will be able to create a message.  Write your own code to complete the steps below.

1. Add 14 to every number.
2. Multiply every number by 18, then subtract 257.
3. Exponentiate every number.  (That is, do e^[number].  You may have to Google how to do this!)
4. Square every number.

**Checkpoint:** Headquarters has informed you that at this stage of decoding, there should be 352 numbers in the secret message that are below 17.  Write code to check if this is true for your data.


5. Turn your one-dimensional array of numbers into a two-dimensional array (matrix) with 5 columns.
6. Separately from your top secret numbers, create an array of all the even numbers between 1 and 382. Name it "evens". That is, "evens" should contain 2, 4, 6, 8 ..., 382.
7. Subtract the "evens" vector from the first column of your secret message matrix.
8. Subtract 100 from all numbers in the 18-24th rows of the 3rd column.
9. Multiply all numbers in the 4th and 5th column by 2.
10. Turn your matrix back into a one dimensional array.


**Checkpoint:** Headquarters has informed you that at this stage of decoding, all numbers in indices 500 and beyond are below 100.  Write code to check if your numbers fulfill this requirement.

11. Take the square root of all numbers in indices 37 to 464.
12. Round all numbers to the nearest whole number.
13. Replace all instances of the number 39 with 20.

**Checkpoint:** Headquarters has informed you that your final message should have 344 even numbers.  Write code to check if your numbers fulfill this requirement.

## Part 3: The secret message!

Run the following to discover the final message!  Google the first line of this message, if you do not recognize it, to see what it is.


In [27]:
''.join(map(str, symbols_df.loc[top_secret-1]['Symbol']))

'\'Twas brillig, and the slithy toves   Did gyre and gimble in the wabe: All mimsy were the borogoves,   And the mome raths outgrabe. "Beware the Jabberwock, my son!   The jaws that bite, the claws that catch! Beware the Jubjub bird, and shun   The frumious Bandersnatch!" He took his vorpal sword in hand;   Long time the manxome foe he sought- So rested he by the Tumtum tree   And stood awhile in thought, And, as in uffish thought he stood,   The Jabberwock, wit: eyes of flame, Came whiffling through the tulgey wood,   And burbled as it came! One, two! One, two! And through and through   The vorpal blade went snicker-snack! He left it dead, and with its head   He went galumphing back. "And hast thou slain the Jabberwock?   Come to my arms, my beamish boy! uEciaaehjjEbamAEoaffhhdAEoaffamACEEEqbEadhisfbbEegEdejEehmzEBwkajEaieffedAEagbEsdbEjfesdmEshkbjEEEoebEdmibEagbEdegafbEegEsdbEkaabFEmffEgegjmEkbibEsdbEahihdhkbjAEEEmgbEsdbEghgbEiasdjEhjsdiaabz'